strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-nuget-layout-provider/src/main/java/org/carlspring/strongbox/nuget/TempNupkgFile.java

```diff
@@ -0,0 +1,235 @@
+     * @throws NugetFormatException
+     *             the stream does not contain the NuGet package or the format
+     *             of the package does not conform to the standard
+     */
+    public TempNupkgFile(InputStream inputStream)
+        throws IOException,
+        NugetFormatException
+    {
+        try
+        {
+            this.file = File.createTempFile("nupkg", "jnuget");
+
+            this.hash = copyDataAndCalculateHash(inputStream, this.file);
+
+            try (final FileInputStream fileInputStream = new FileInputStream(file))
+            {
+                this.nuspec = loadNuspec(fileInputStream);
+            }
+        }
+        catch (NoSuchAlgorithmException ex)
+        {
+            throw new NugetFormatException("Unable to calculate hash of package", ex);
+        }
+    }
+
+    /**
+     * Extract specification file from stream with NuPkg package
+     *
+     * @param package
+     *            stream stream package
+     * @return specification file
+     * @throws IOException
+     *             read error
+     * @throws NugetFormatException
+     *             XML in the package archive does not conform to the NuGet
+     *             specification
+     */
+    private static final Nuspec loadNuspec(InputStream packageStream)
+        throws IOException,
+        NugetFormatException
+    {
+        try (ZipInputStream zipInputStream = new ZipInputStream(packageStream);)
+        {
+            ZipEntry entry;
+            do
+            {
+                entry = zipInputStream.getNextEntry();
+            } while (entry != null && !isNuspecZipEntry(entry));
+
+            if (entry == null)
+            {
+                return null;
+            }
+
+            return Nuspec.parse(zipInputStream);
+        }
+    }
+
+    /**
+     * ZIP attachment is Nuspec XML specification
+     *
+     * @param entry
+     *            zip attachment
+     * @return true if the attachment matches the attachment with the
+     *         specification
+     */
+    private static boolean isNuspecZipEntry(ZipEntry entry)
+    {
+        return !entry.isDirectory() && entry.getName().endsWith(Nuspec.DEFAULT_FILE_EXTENSION);
+    }
+
+    /**
+     * Creates a temporary file based on the stream.
+     *
+     * @param inputStream
+     *            data stream
+     * @param targetFile
+     *            file to copy the package to
+     * @return data file
+     * @throws IOException
+     *             read / write error
+     * @throws NoSuchAlgorithmException
+     *             the system does not have an algorithm for calculating the
+     *             value of HASH
+     */
+    private static String copyDataAndCalculateHash(InputStream inputStream,
+                                                   File targetFile)
+        throws IOException,
+        NoSuchAlgorithmException
+    {
+        MessageDigest messageDigest = MessageDigest.getInstance(ALGORITHM_NAME);
```

Comment:
This code is actually quite peculiar...

@sbespalov : This code should be re-worked right? (@kalski : This could be a follow-up task, if you like).

We have a class called `org.carlspring.commons.io.MultipleDigestInputStream` which should perform better. The idea is that while you're reading the `InputStream` the actual digest is being updated in real-time, so when you're done reading, you can just calculate it at the end. This is a lot more efficient, because it doesn't require an additional re-read. The places which are referencing this method would have to be re-worked to use the `MultipleDigestInputStream`; (in addition, as the name infers, you can calculate more than one checksum this way).

Reply:
@carlspring probably we can consider to use `MultipleDigestInputStream` here but I suggest to leave this as followup tasks.

Reply:
@kalski ,
Would you mind creating a follow up task for this on our issue tracker?
Thanks!

Reply:
@carlspring @sbespalov Note that `TempNupkgFile` checksum calculation already decorates `InputStream` with `DigestInputStream` while creating the temp file to be later read with `ZipInputStream`. 

I understand that the most efficient implementation would read the nuspec out of the zip and calculate zip's digest at the same time but I tend to agree this should be carefully implemented and tested in a separate task. 
I also noticed that we have `LayoutInputStream` which in a way copies `MultipleDigestInputStream` logic so I guess we could also use it instead. Can you advise or maybe create this follow up issue?

Reply:
@sbespalov ,

Any thoughts?

Reply:
I will create an issue

Reply:
I will create an issue